File: kestrel-kotest/src/main/kotlin/io/github/matthewjones372/kestrel/kotest/Kestrel.kt
package io.github.matthewjones372.kestrel.kotest

import io.github.matthewjones372.kestrel.Engine
import io.github.matthewjones372.kestrel.Progress
import io.github.matthewjones372.kestrel.engine.Kestrel
import io.github.matthewjones372.kestrel.engine.exclusive
import kotlin.coroutines.AbstractCoroutineContextElement
import kotlin.coroutines.CoroutineContext
import kotlin.coroutines.coroutineContext

/**
 * The runner for the test this is called from.
 *
 * A Kotest test body is a coroutine, so the runner rides in the coroutine
 * context rather than in a thread local — a test that switches dispatchers
 * would lose a thread local halfway through and start recording into nothing.
 *
 * Called outside [KestrelExtension] it still works, handing back a runner of
 * its own: a spec should not have to register anything to run a simulation.
 *
 * Silent either way: a Kotest report is somebody else's output, and a run that
 * prints a line every five seconds into it is noise a reader has to scroll past
 * to reach the failure.
 */
suspend fun kestrel(): Kestrel = coroutineContext[Running]?.kestrel ?: Kestrel(Progress.silent)

/**
 * A runner for the test this is called from that sends through [engine], as
 * silent as the one that takes no argument.
 *
 * An overload rather than a context element or a spec-level extension:
 * reading an element back is stdlib, but installing one is `withContext` from
 * kotlinx-coroutines, and registering an extension is Kotest, neither of which
 * this module carries at runtime. An engine is a value a spec can hold in a
 * `val` and hand to each test that wants it, which is what the other two would
 * have arranged more slowly.
 *
 * The machine is taken for a run the same way the default takes it, so two
 * specs running at once measure the machine one after the other rather than
 * measuring each other.
 */
fun kestrel(engine: Engine): Kestrel = Kestrel(engine.exclusive(), Progress.silent)

internal class Running(val kestrel: Kestrel) : AbstractCoroutineContextElement(Running) {
    companion object Key : CoroutineContext.Key<Running>
}

File: kestrel-zio-test/src/main/scala/io/github/matthewjones372/kestrel/ziotest/Kestrel.scala
package io.github.matthewjones372.kestrel.ziotest

import io.github.matthewjones372.kestrel.Engine
import io.github.matthewjones372.kestrel.Progress
import io.github.matthewjones372.kestrel.RunResult
import io.github.matthewjones372.kestrel.Simulation
import io.github.matthewjones372.kestrel.engine.ExclusiveKt
import zio.Task
import zio.ZIO
import io.github.matthewjones372.kestrel.engine.Kestrel as Runner

/**
 * The runner for the test this is called from.
 *
 * Blocking rather than compute, and that is the whole module. The call blocks
 * its thread for the length of the run while the engine sends on virtual
 * threads; on ZIO's compute pool that is a starved runtime, and a starved
 * runtime is a scheduler this tool would then measure and report as the
 * target's latency.
 *
 * Nothing here puts a fiber between the departure clock and the socket: a step
 * body is an `Action` and stays one. What a run measures is the wall clock, so
 * `TestClock` cannot move it and must not.
 *
 * Silent either way: a zio-test report is somebody else's output, and a run
 * that prints a line every five seconds into it is noise a reader has to
 * scroll past to reach the failure.
 */
object kestrel:

  def run(simulation: Simulation): Task[RunResult] =
    ZIO.attemptBlocking(Runner(Progress.Companion.getSilent).run(simulation))

  /**
   * The same run, sent by [[on]] rather than by the default engine.
   *
   * The machine is taken for the run the way the default takes it, so two
   * tests under `TestAspect.parallel` measure it one after the other rather
   * than measuring each other.
   */
  def run(simulation: Simulation, on: Engine): Task[RunResult] =
    val silent = Progress.Companion.getSilent
    ZIO.attemptBlocking(Runner(ExclusiveKt.exclusive(on, silent), silent).run(simulation))
